pydantic/_internal/_discriminated_union.py

```diff
@@ -181,10 +181,11 @@ def _apply_to_root(self, schema: core_schema.CoreSchema) -> core_schema.CoreSche
             return definitions_wrapper
 
         if schema['type'] != 'union':
-            raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
-
-        if len(schema['choices']) < 2:
-            raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+            # If the schema is not a union, it probably means it just had a single member and
+            # was flattened by pydantic_core.
```

Comment:
As per other comment I think it's the `Union` implementation, not `pydantic_core`, responsible for the flattening.

Reply:
When I get here in the debugger, I actually see my type is `tagged-union`

Reply:
@davidhewitt there is also a thing in pydantic-core that flattens unions. I think in principle you could get a `core_schema.union_schema([<one_choice>])`, that this would handle. I agree it's not likely to be hit currently.